lib/url.js

```diff
@@ -10,60 +10,73 @@ exports.format = urlFormat;
 exports.Url = Url;
 
 function Url() {
-  this.protocol = null;
-  this.slashes = null;
+  // For more efficient internal representation and laziness.
+  // The non-underscore versions of these properties are accessor functions
+  // defined on the prototype.
+  this._protocol = null;
+  this._href = '';
+  this._port = -1;
+  this._query = null;
+  this._prependSlash = false;
   this.auth = null;
+  this.slashes = null;
   this.host = null;
-  this.port = null;
   this.hostname = null;
   this.hash = null;
   this.search = null;
-  this.query = null;
   this.pathname = null;
-  this.path = null;
-  this.href = null;
 }
 
 // Reference: RFC 3986, RFC 1808, RFC 2396
 
-// define these here so at least they only have to be
-// compiled once on the first module load.
-const protocolPattern = /^([a-z0-9.+-]+:)/i;
-const portPattern = /:[0-9]*$/;
-
-// Special case for a simple path URL
-const simplePathPattern = /^(\/\/?(?!\/)[^\?\s]*)(\?[^\s]*)?$/;
+const _protocolCharacters = makeAsciiTable([
+  [0x61, 0x7A] /*a-z*/,
+  [0x41, 0x5A] /*A-Z*/,
+  0x2E /*'.'*/, 0x2B /*'+'*/, 0x2D /*'-'*/
+]);
 
 // RFC 2396: characters reserved for delimiting URLs.
 // We actually just auto-escape these.
-const delims = ['<', '>', '"', '`', ' ', '\r', '\n', '\t'];
-
 // RFC 2396: characters not allowed for various reasons.
-const unwise = ['{', '}', '|', '\\', '^', '`'].concat(delims);
-
 // Allowed by RFCs, but cause of XSS attacks.  Always escape these.
-const autoEscape = ['\''].concat(unwise);
+const _autoEscape = [
+  '<',
+  '>',
+  '\'',
+  '`',
+  ' ',
+  '\r',
+  '\n',
+  '\t',
+  '{',
+  '}',
+  '|',
+  '\\',
+  '^',
+  '`',
+  '"'
+];
+
+const _autoEscapeMap = new Array(128);
+
+for (var i = 0, len = _autoEscapeMap.length; i < len; ++i) {
+  _autoEscapeMap[i] = '';
+}
 
-// Characters that are never ever allowed in a hostname.
-// Note that any invalid chars are also handled, but these
-// are the ones that are *expected* to be seen, so we fast-path them.
-const nonHostChars = ['%', '/', '?', ';', '#'].concat(autoEscape);
-const hostEndingChars = ['/', '?', '#'];
-const hostnameMaxLen = 255;
-const hostnamePartPattern = /^[+a-z0-9A-Z_-]{0,63}$/;
-const hostnamePartStart = /^([+a-z0-9A-Z_-]{0,63})(.*)$/;
-// protocols that can allow "unsafe" and "unwise" chars.
-const unsafeProtocol = {
-  'javascript': true,
-  'javascript:': true
-};
-// protocols that never have a hostname.
-const hostlessProtocol = {
-  'javascript': true,
-  'javascript:': true
-};
-// protocols that always contain a // bit.
-const slashedProtocol = {
+for (var i = 0, len = _autoEscape.length; i < len; ++i) {
+  var c = _autoEscape[i];
```

Comment:
I don't know performance effect. 
But I think `let` / `const` declaration is more suitable in here.
no hoisting, if we declare `c` variable as `var`, we can use `c` out of this for-loop.
however, if we declare `c` as `let`, we can not use `c` out of the for-loop (Block Scope).

And V8 is implementing [strong mode](https://developers.google.com/v8/experiments#strong) experimentally, if strong mode will be on, v8 will output warnings like `In strong mode, 'var' is deprecated, use 'let' or 'const' instead"`.